src/utils/DockerInstallStatusProvider.ts

```diff
@@ -0,0 +1,36 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AsyncLazy } from "./lazy";
+import { execAsync } from "./spawnAsync";
+
+export class DockerInstallStatusProvider {
+    private static instance: DockerInstallStatusProvider;
+    private maxCacheTime: number = 5000;
+    private isDockerInstalledLazy: AsyncLazy<boolean>;
+
+    private constructor() {
+        this.isDockerInstalledLazy = new AsyncLazy<boolean>(async () => {
+            try {
+                await execAsync('docker -v');
+                return true; // As long as the docker command did't throw exception, assume it is installed.
+            } catch (error) {
+                return false; // docker not installed
+            }
+        }, this.maxCacheTime);
+    }
+
+    public static getInstance(): DockerInstallStatusProvider {
```

Comment:
We use a simpler pattern for singletons in the Docker extension codebase, e.g. see https://github.com/microsoft/vscode-docker/blob/master/src/utils/dockerContextManager.ts#L241